### PR TITLE
tests: cover txn.CheckTxnAuth logic with unit tests

### PR DESCRIPTION
**Pull request by vianamjr**
Related issue: #15707

Increase test coverage.

This cover 100% the function txn.CheckTxnAuth

### CODE REVIEW

**Comment by vianamjr** [comment]
@serathius is there something else for now that I need to do? (It's my first contribution)

File: server/etcdserver/txn/txn_test.go [hunk 1]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name       string
+		txnRequest *pb.TxnRequest
+		err        error
+	}{
+		{
+			name: "Unauthorize out of range compare",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:162]:
```suggestion
			name: "Out of range compare is unathorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 2]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+					{
+						Key: []byte("boo"),
+					},
+				},
+				Success: []*pb.RequestOp{},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:174]:
```suggestion
			name: "Nil request range is always authorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 3]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request range in range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:187]:
```suggestion
			name: "Range request in range is authorised",
```

File: server/etcdserver/txn/txn_test.go [hunk 4]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorize request range out of range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:203]:
```suggestion
			name: "Range request out of range is unauthorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 5]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request put",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:219]:
```suggestion
			name: "Nil Put request is authorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 6]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request pur in range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:232]:
```suggestion
			name: "Put request in range in authorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 7]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+								Key: []byte("foo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorized request pur in range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:247]:
```suggestion
			name: "Put request out of range is unauthorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 8]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+								Key: []byte("boo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil delete range",
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:262]:
```suggestion
			name: "Nil delete request is authorized",
```

File: server/etcdserver/txn/txn_test.go [hunk 9]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+						Key: []byte("foo"),
+					},
+				},
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestDeleteRange{
+							RequestDeleteRange: &pb.DeleteRangeRequest{
+								Key:      []byte("foo"),
+								RangeEnd: []byte("zoo"),
+								PrevKv:   true,
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:288]: Any reason to set `PrevKv` here?

File: server/etcdserver/txn/txn_test.go [hunk 10]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize delete range in range compare and rerquest",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("foo"),
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:279]: Is there any reason that you don't test compare out of range?

File: server/etcdserver/txn/txn_test.go [hunk 11]
```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
...
+		},
+		{
+			name: "Unauthorize delete range out of range keys",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("foo"),
+					},
+				},
+				Success: []*pb.RequestOp{
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:304]: Is there any reason you test only Txn.Success and not Txn.Failure? Either of them can execute depending on compare result.

File: server/etcdserver/txn/txn_test.go (hunk 9 above)
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:288]: Any reason you test DeleteRangeRequest with Compare? Could you test Compare idependently?

File: server/etcdserver/txn/txn_test.go (hunk 11 above)
Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go:433]: I added the tests

File: server/etcdserver/txn/txn_test.go (hunk 9 above)
Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go:394]: I separated it.

File: server/etcdserver/txn/txn_test.go (hunk 10 above)
Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go]: I added.

**Comment by vianamjr** [comment]
Not sure If I coverage all your comments! Thanks for the review. Please let me know if something else is needed.

File: server/etcdserver/txn/txn_test.go [hunk 12]
```diff
@@ -94,3 +98,429 @@ func TestWriteTxnPanic(t *testing.T) {
...
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:200]: Those tests have needles repetition and bloat. Have you thought about refactoring some parts?

For example creating variable for repeating parts
```
inRangeRequest := RequestOp{
	Request: &pb.RequestOp_RequestRange{
		RequestRange: &pb.RangeRequest{
			Key:      []byte("foo"),
			RangeEnd: []byte("zoo"),
		},
	},
},
```
Then the test could be just
```
{
	name: "Range request in range is authorized",
	txnRequest: &pb.TxnRequest{
		Success: []*pb.RequestOp{inRangeRequest},
		Failure: []*pb.RequestOp{inRangeRequest},
	},
	err: nil,
},
```

Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go:274]: Makes sense.

File: server/etcdserver/txn/txn_test.go [hunk 13]
```diff
@@ -94,3 +98,289 @@ func TestWriteTxnPanic(t *testing.T) {
 
 	assert.Panics(t, func() { Txn(ctx, zaptest.NewLogger(t), txn, false, s, &lease.FakeLessor{}) }, "Expected panic in Txn with writes")
 }
+
+func TestCheckTxnAuth(t *testing.T) {
```
Comment (serathius) [inline server/etcdserver/txn/txn_test.go:102]: Last thing, the test function is just too long. Could you:
* Move auth setup to dedicated function
* Move variables like `inRangeCompare` out of the test.

For example create function
```
func setupAuth(t *testing.T, be backend.Backend) {
  simpleTokenTTLDefault := 300 * time.Second
	tokenTypeSimple := "simple"
	dummyIndexWaiter := func(index uint64) <-chan struct{} {
		ch := make(chan struct{}, 1)
		go func() {
			ch <- struct{}{}
		}()
		return ch
	}

	tp, _ := auth.NewTokenProvider(zaptest.NewLogger(t), tokenTypeSimple, dummyIndexWaiter, simpleTokenTTLDefault)

	as := auth.NewAuthStore(lg, schema.NewAuthBackend(lg, be), tp, 4)

	// create "root" user and "foo" user with limited range
	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "rw"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{
		Name: "rw",
		Perm: &authpb.Permission{
			PermType: authpb.READWRITE,
			Key:      []byte("foo"),
			RangeEnd: []byte("zoo"),
		},
	}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "root", Password: "foo"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "foo", Password: "foo"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "root", Role: "root"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "foo", Role: "rw"}); err != nil {
		t.Fatal(err)
	}

	if err := as.AuthEnable(); err != nil {
			t.Fatal(err)
	}
}

In the end test function should look like 
```
func TestCheckTxnAuth(t *testing.T) {
	lg := zaptest.NewLogger(t)

	be, _ := betesting.NewDefaultTmpBackend(t)
	defer betesting.Close(t, be)
	setupAuth(t, be)
		tests := []struct {
		name       string
		txnRequest *pb.TxnRequest
		err        error
	}{
		{
			name: "Out of range compare is unauthorized",
			txnRequest: &pb.TxnRequest{
				Compare: []*pb.Compare{outOfRangeCompare},
			},
			err: auth.ErrPermissionDenied,
		},
	...
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := CheckTxnAuth(as, &auth.AuthInfo{Username: "foo", Revision: 8}, tt.txnRequest)
			if err != tt.err {
				t.Errorf("expected error to be: %v; got: %v", tt.err, err)
			}
		})
	}
}

}
```

Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go:103]: Perfect!

**Review by serathius — APPROVED** [review]

**Comment by serathius** [comment]
cc @ahrtr

**Comment by serathius** [comment]
cc @lavacat

**Comment by vianamjr** [comment]
As it's is my first contribution, is there anything else necessary from my side?

File: server/etcdserver/txn/txn_test.go [hunk 14]
```diff
@@ -94,3 +99,287 @@ func TestWriteTxnPanic(t *testing.T) {
...
+			err: auth.ErrPermissionDenied,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := CheckTxnAuth(as, &auth.AuthInfo{Username: "foo", Revision: 8}, tt.txnRequest)
+			if err != tt.err {
+				t.Errorf("expected error to be: %v; got: %v", tt.err, err)
+			}
```
Comment (ahrtr) [inline server/etcdserver/txn/txn_test.go:243]: minor comment: 
```suggestion
			err := CheckTxnAuth(as, &auth.AuthInfo{Username: "foo", Revision: 8}, tt.txnRequest)
			assert.Equal(t, tt.err, err)
```

**Review by ahrtr — APPROVED** [review]
LGTM with a couple of minor comments.

Thank you @vianamjr

**Comment by ahrtr** [comment]
> As it's is my first contribution, is there anything else necessary from my side?

Once you resolve the minor comments above, I will merge this PR.

File: server/etcdserver/txn/txn_test.go [hunk 15]
```diff
@@ -265,10 +264,10 @@ func setupAuth(t *testing.T, be backend.Backend) auth.AuthStore {
 
 	// create "root" user and "foo" user with limited range
 	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"}); err != nil {
-		t.Fatal(err)
+		require.NoError(t, err)
```
Comment (ahrtr) [inline server/etcdserver/txn/txn_test.go:267]: It isn't correct.

Please change 
```
	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"}); err != nil {
		t.Fatal(err)
		require.NoError(t, err)
	}
```

to 

```
	err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"})
	require.NoError(t, err)
```

Comment (vianamjr) [reply server/etcdserver/txn/txn_test.go:267]: Now I see, thanks.

**Pull request merged**